clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaClientFactory.java

```diff
@@ -0,0 +1,190 @@
+        return resultPromise.future();
+    }
+
+    private Instant getRetriesTimeLimit(final Duration retriesTimeout) {
+        if (retriesTimeout == null || retriesTimeout.isNegative()) {
+            return Instant.MAX;
+        }
+        return Instant.now(clock).plus(retriesTimeout);
+    }
+
+    private <T> Future<T> createClientWithRetries(
+            final String bootstrapServersConfig,
+            final Supplier<T> clientSupplier,
+            final Instant retriesTimeLimit,
+            final Promise<T> resultPromise) {
+        try {
+            resultPromise.complete(clientSupplier.get());
+        } catch (final Exception e) {
+            // perform retry in case bootstrap URLs are not resolvable ("No resolvable bootstrap urls given in bootstrap.servers")
+            // (see org.apache.kafka.clients.ClientUtils#parseAndValidateAddresses)
+            if (e instanceof KafkaException
+                    && e.getCause() instanceof ConfigException && e.getCause().getMessage() != null
+                    && e.getCause().getMessage().contains(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG)
+                    && containsValidServerEntries(bootstrapServersConfig)) {
+                if (Instant.now(clock).isAfter(retriesTimeLimit)) {
+                    LOG.warn("error creating Kafka client (no further attempts will be done, timeout for retries reached): {}",
+                            e.getCause().getMessage());
+                    resultPromise.fail(e);
+                } else {
+                    LOG.debug("error creating Kafka client, will retry in {}ms: {}", CLIENT_CREATION_RETRY_DELAY_MILLIS,
+                            e.getCause().getMessage());
+                    vertx.setTimer(CLIENT_CREATION_RETRY_DELAY_MILLIS, tid -> {
+                        createClientWithRetries(bootstrapServersConfig, clientSupplier, retriesTimeLimit, resultPromise);
+                    });
+                }
+            } else {
+                resultPromise.fail(e);
+            }
+        }
+        return resultPromise.future();
+    }
+
+    private static boolean containsValidServerEntries(final String bootstrapServersConfig) {
+        final List<String> urlList = Optional.ofNullable(bootstrapServersConfig)
+                .map(serversString -> {
+                    final String trimmed = serversString.trim();
+                    if (trimmed.isEmpty()) {
+                        return List.<String> of();
+                    }
+                    return Arrays.asList(COMMA_WITH_WHITESPACE.split(trimmed, -1));
+                }).orElseGet(List::of);
+        return !urlList.isEmpty() && urlList.stream().allMatch(KafkaClientFactory::containsHostAndPort);
+    }
+
+    private static boolean containsHostAndPort(final String url) {
+        try {
+            return Utils.getHost(url) != null && Utils.getPort(url) != null;
+        } catch (final IllegalArgumentException e) {
+            return false;
```

Comment:
I don't think it's needed here. In case host or port are invalid, the client creation will fail with a corresponding exception (`Invalid url in bootstrap.servers: [url]`). And any kind of client creation exception should be logged in the invoking code anyway.